src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -542,34 +615,39 @@ module Make (Inputs : Inputs_intf.S) = struct
     (* transfer state from mask to parent; flush local state *)
     let commit t =
       assert_is_attached t ;
+      let parent = get_parent t in
       let old_root_hash = merkle_root t in
-      let account_data = Map.to_alist !(t.accounts) in
-      Base.set_batch (get_parent t) account_data ;
-      t.accounts := Location_binable.Map.empty ;
-      t.hashes := Addr.Map.empty ;
+      let account_data = Map.to_alist t.maps.accounts in
+      Base.set_batch parent account_data ;
+      t.maps.accounts <- Location_binable.Map.empty ;
+      t.maps.hashes <- Addr.Map.empty ;
+      (* TODO why only 2/4 maps are updated ? *)
       Debug_assert.debug_assert (fun () ->
           [%test_result: Hash.t]
             ~message:
               "Parent merkle root after committing should be the same as the \
                old one in the mask"
-            ~expect:old_root_hash
-            (Base.merkle_root (get_parent t)) ;
+            ~expect:old_root_hash (Base.merkle_root parent) ;
           [%test_result: Hash.t]
             ~message:"Merkle root of the mask should delegate to the parent now"
-            ~expect:(merkle_root t)
-            (Base.merkle_root (get_parent t)) )
+            ~expect:(merkle_root t) (Base.merkle_root parent) )
 
     (* copy tables in t; use same parent *)
     let copy t =
+      let detached_parent_signal = Async.Ivar.create () in
       { uuid = Uuid_unix.create ()
       ; parent = Ok (get_parent t)
-      ; detached_parent_signal = Async.Ivar.create ()
-      ; accounts = ref !(t.accounts)
-      ; token_owners = ref !(t.token_owners)
-      ; locations = ref !(t.locations)
-      ; hashes = ref !(t.hashes)
+      ; detached_parent_signal
       ; current_location = t.current_location
       ; depth = t.depth
+      ; maps = maps_copy t.maps
+      ; accumulated =
+          Option.map t.accumulated ~f:(fun acc ->
+              { base = acc.base
+              ; detached_next_signal = detached_parent_signal
```

Comment:
Yeah it seems that the logic here is meant to be that we carry the `detached_next_signal` through until we hit a layer where we trigger a rotation. Depending on how we use `copy` in the code, this could lead to the accumulator never being rotated, causing a memory leak.

Reply:
This is a bug indeed
Initial implementation got it right, then in 3dba63a0f599 I changed it to what it is now, but this was a wrong change